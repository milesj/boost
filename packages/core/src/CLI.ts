/**
 * @copyright   2017-2018, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

// eslint-disable-next-line import/no-extraneous-dependencies
import { Argv } from 'yargs';
import Tool from './Tool';

export default class CLI {
  static registerGlobalOptions(app: Argv, tool: Tool) {
    app
      .option('config', {
        default: '',
        description: tool.msg('app:cliOptionConfig'),
        string: true,
      })
      .option('debug', {
        boolean: true,
        default: false,
        description: tool.msg('app:cliOptionDebug'),
      })
      .option('locale', {
        default: '',
        description: tool.msg('app:cliOptionLocale'),
        string: true,
      })
      .option('output', {
        default: 3,
        description: tool.msg('app:cliOptionOutput'),
        number: true,
      })
      .option('silent', {
        boolean: true,
        default: false,
        description: tool.msg('app:cliOptionSilent'),
      })
      .option('theme', {
        default: 'default',
        description: tool.msg('app:cliOptionTheme'),
        string: true,
      });
  }
}
