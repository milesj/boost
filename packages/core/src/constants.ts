import { Status } from './types';

export const APP_NAME_PATTERN = /^[a-z]{1}[-a-z0-9]+[a-z]{1}$/u;
export const CONFIG_NAME_PATTERN = /^[a-z]{1}[a-zA-Z0-9]+$/u;
export const EVENT_NAME_PATTERN = /^[-a-z.]+$/u;
export const MODULE_NAME_PATTERN = /^(@[-a-z]+\/)?[-a-z]+$/u;
export const PLUGIN_NAME_PATTERN = /^([a-z]+):[-a-z]+$/u;

export const STATUS_PENDING: Status = 'pending';
export const STATUS_RUNNING: Status = 'running';
export const STATUS_SKIPPED: Status = 'skipped';
export const STATUS_PASSED: Status = 'passed';
export const STATUS_FAILED: Status = 'failed';

// Chalk colors
// yellow = app name, module name
// cyan = file
// magenta = plugin name
