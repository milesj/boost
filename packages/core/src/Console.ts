/**
 * @copyright   2017-2019, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import exit from 'exit';
import cliSize from 'term-size';
import ansiEscapes from 'ansi-escapes';
import Emitter from './Emitter';
import Tool from './Tool';
import { Debugger } from './types';
import Output from './Output';

// 16 FPS (60 FPS is actually too fast as it tears)
export const FPS_RATE = 62.5;

// Bind our writable streams for easy access
export const BOUND_WRITERS = {
  stderr: process.stderr.write.bind(process.stderr),
  stdout: process.stdout.write.bind(process.stdout),
};

export const WRAPPED_STREAMS = {
  stderr: false,
  stdout: false,
};

export type StreamType = 'stderr' | 'stdout';

export interface ConsoleState {
  disabled: boolean;
  final: boolean;
  started: boolean;
  stopped: boolean;
}

export default class Console extends Emitter {
  bufferedStreams: (() => void)[] = [];

  debug: Debugger;

  errorLogs: string[] = [];

  logs: string[] = [];

  outputQueue: Output[] = [];

  tool: Tool<any>;

  protected renderTimer: NodeJS.Timer | null = null;

  protected restoreCursorOnExit: boolean = false;

  protected state: ConsoleState = {
    disabled: false,
    final: false,
    started: false,
    stopped: false,
  };

  private writers: typeof BOUND_WRITERS;

  constructor(tool: Tool<any>, /* test only */ testWriters: typeof BOUND_WRITERS = BOUND_WRITERS) {
    super();

    this.debug = tool.createDebugger('console');
    this.tool = tool;
    this.writers = testWriters;

    // istanbul ignore next
    if (process.env.NODE_ENV !== 'test') {
      process
        .on('SIGINT', this.handleSignal)
        .on('SIGTERM', this.handleSignal)
        .on('uncaughtException', this.handleFailure)
        .on('unhandledRejection', this.handleFailure);
    }
  }

  /**
   * Disable the render loop entirely.
   */
  disable(): this {
    this.state.disabled = true;

    return this;
  }

  /**
   * Display a footer after all output.
   */
  displayFooter() {
    const { footer } = this.tool.options;

    if (footer) {
      this.out(footer, 1);
    }
  }

  /**
   * Display a header before all output.
   */
  displayHeader() {
    const { header } = this.tool.options;

    if (header) {
      this.out(header, 1);
    }
  }

  /**
   * Enable the render loop.
   */
  enable(): this {
    this.state.disabled = false;

    return this;
  }

  /**
   * Write a message to `stderr` with optional trailing newline(s).
   */
  err(message: string, nl: number = 0): this {
    if (!this.isSilent()) {
      this.writers.stderr(message + '\n'.repeat(nl));
    }

    return this;
  }

  /**
   * Flush buffered stream output.
   */
  flushBufferedStreams(): this {
    this.bufferedStreams.forEach(buffer => {
      buffer();
    });

    return this;
  }

  /**
   * Flush the top output block in the queue.
   */
  flushOutputQueue(): this {
    const outputs = this.outputQueue.filter((out, i) => i === 0 || out.isConcurrent());

    // Erase the previous output
    outputs.forEach(output => {
      output.erasePrevious();
    });

    // Write buffered output
    this.flushBufferedStreams();

    // Write the next output
    outputs.forEach(output => {
      output.render();
    });

    // Remove completed outputs
    this.outputQueue = this.outputQueue.filter(out => !out.isComplete());

    // Stop the render loop once the queue is empty
    if (this.isEmptyQueue()) {
      this.stopRenderLoop();
    } else {
      this.startRenderLoop();
    }

    return this;
  }

  /**
   * Handle uncaught exceptions and unhandled rejections that bubble up.
   */
  handleFailure = (error: Error) => {
    this.start();
    this.debug('Uncaught exception or unresolved promise handled');
    this.stop(error);

    exit(2);
  };

  /**
   * Handle SIGINT and SIGTERM interruptions.
   */
  handleSignal = () => {
    this.start();
    this.debug('SIGINT or SIGTERM handled');
    this.stop(new Error(this.tool.msg('errors:processTerminated')));

    exit(2);
  };

  /**
   * Hide the console cursor.
   */
  hideCursor(): this {
    this.out(ansiEscapes.cursorHide);

    if (!this.restoreCursorOnExit && !this.isSilent()) {
      this.restoreCursorOnExit = true;

      // istanbul ignore next
      process.on('exit', () => {
        process.stdout.write(ansiEscapes.cursorShow);
      });
    }

    return this;
  }

  /**
   * Return true if the render loop has been disabled.
   */
  isDisabled(): boolean {
    return this.state.disabled;
  }

  /**
   * Return true if the output queue is empty.
   */
  isEmptyQueue(): boolean {
    return this.outputQueue.length === 0;
  }

  /**
   * Return true if the final render.
   */
  isFinalRender(): boolean {
    return this.state.final;
  }

  /**
   * Return true if the there should be no output.
   */
  isSilent(): boolean {
    return this.tool.config.silent;
  }

  /**
   * Return true if the defined stream has been wrapped by the console layer.
   */
  isStreamWrapped(type: StreamType): boolean {
    return !!WRAPPED_STREAMS[type];
  }

  /**
   * Log a message to display on success during the final render.
   */
  log(message: string): this {
    this.logs.push(message);

    return this;
  }

  /**
   * Log a live message to display during the rendering process.
   */
  logLive(message: string): this {
    // Write to the wrapped buffer
    process.stdout.write(message);

    return this;
  }

  /**
   * Log an error message to display on failure during the final render.
   */
  logError(message: string): this {
    this.errorLogs.push(message);

    return this;
  }

  /**
   * Write a message to `stdout` with optional trailing newline(s).
   */
  out(message: string, nl: number = 0): this {
    if (!this.isSilent()) {
      this.writers.stdout(message + '\n'.repeat(nl));
    }

    return this;
  }

  /**
   * Enqueue a block of output to be rendered.
   */
  render(output: Output): this {
    if (this.isDisabled()) {
      throw new Error(
        'Output cannot be enqueued as the render loop has been disabled. This is usually caused by conflicting reporters.',
      );
    }

    if (!this.outputQueue.includes(output)) {
      this.outputQueue.push(output);
    }

    // Only run the render loop when output is enqueued
    this.startRenderLoop();

    return this;
  }

  /**
   * Handle the final rendering of all output before stopping.
   */
  renderFinalOutput(error: Error | null) {
    this.debug('Rendering final console output');
    this.state.final = true;

    // Mark all output as final
    this.outputQueue.forEach(output => {
      output.enqueue(true);
    });

    // Recursively render the remaining output
    while (!this.isEmptyQueue()) {
      this.flushOutputQueue();
    }

    // Stop the render loop
    this.stopRenderLoop();

    if (error) {
      if (this.errorLogs.length > 0) {
        this.err(`\n${this.errorLogs.join('\n')}\n`);
      }

      this.emit('error', [error]);
    } else {
      if (this.logs.length > 0) {
        this.out(`\n${this.logs.join('\n')}\n`);
      }

      this.displayFooter();
    }

    // Flush any stream output that still exists
    this.flushBufferedStreams();

    // Show the cursor incase it has been hidden
    this.out(ansiEscapes.cursorShow);
  }

  /**
   * Reset the cursor back to the bottom of the console.
   */
  resetCursor(): this {
    this.out(ansiEscapes.cursorTo(0, cliSize().rows));

    return this;
  }

  /**
   * Show the console cursor.
   */
  showCursor(): this {
    this.restoreCursorOnExit = false;
    this.out(ansiEscapes.cursorShow);

    return this;
  }

  /**
   * Start the console by wrapping streams and buffering output.
   */
  start(args: any[] = []): this {
    if (this.state.started) {
      return this;
    }

    this.debug('Starting console render loop');
    this.emit('start', args);
    this.wrapStreams();
    this.displayHeader();
    this.state.started = true;

    return this;
  }

  /**
   * Automatically render the console in a timeout loop at 16 FPS.
   */
  startRenderLoop() {
    if (this.isSilent() || this.isDisabled()) {
      return;
    }

    this.renderTimer = setTimeout(() => {
      this.flushOutputQueue();
    }, FPS_RATE);
  }

  /**
   * Stop the console rendering process.
   */
  stop(error: Error | null = null) {
    if (this.state.stopped) {
      return;
    }

    if (error) {
      this.debug('Stopping console with an error');
      this.errorLogs.push(...this.logs);
      this.logs = [];
    } else {
      this.debug('Stopping console render loop');
    }

    this.renderFinalOutput(error);
    this.unwrapStreams();
    this.emit('stop', [error]);
    this.state.stopped = true;
    this.state.started = false;
  }

  /**
   * Stop the background render loop.
   */
  stopRenderLoop() {
    if (this.renderTimer) {
      clearTimeout(this.renderTimer);
      this.renderTimer = null;
    }
  }

  /**
   * Unwrap the native console and reset it back to normal.
   */
  unwrapStreams() {
    if (this.isSilent() || process.env.NODE_ENV === 'test') {
      return;
    }

    ['stderr', 'stdout'].forEach(key => {
      const name = key as StreamType;

      if (!this.isStreamWrapped(name)) {
        return;
      }

      this.debug('Unwrapping `%s` stream', name);

      process[name].write = this.writers[name] as any;

      WRAPPED_STREAMS[name] = false;
    });
  }

  /**
   * Wrap the `stdout` and `stderr` streams and buffer the output as
   * to not collide with our render loop.
   */
  wrapStreams() {
    if (this.isSilent() || this.isDisabled() || process.env.NODE_ENV === 'test') {
      return;
    }

    ['stderr', 'stdout'].forEach(key => {
      const name = key as StreamType;
      const stream = process[name];
      let buffer = '';

      if (this.isStreamWrapped(name)) {
        return;
      }

      this.debug('Wrapping `%s` stream', name);

      this.bufferedStreams.push(() => {
        if (stream.isTTY && buffer) {
          this.writers[name](buffer);
        }

        buffer = '';
      });

      stream.write = (chunk: string) => {
        // No output, display immediately
        if (this.isEmptyQueue()) {
          this.writers[name](chunk);
        } else {
          buffer += String(chunk);
        }

        return true;
      };

      WRAPPED_STREAMS[name] = true;
    });
  }
}
