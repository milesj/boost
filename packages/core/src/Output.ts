/**
 * @copyright   2019, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import cliSize from 'term-size';
import ansiEscapes from 'ansi-escapes';
import Console from './Console';

export type Renderer = () => string;

export default class Output {
  protected completed: boolean = false;

  protected console: Console;

  protected final: boolean = false;

  protected previousHeight: number = 0;

  protected renderer: Renderer;

  constructor(cli: Console, renderer: Renderer) {
    if (typeof renderer !== 'function') {
      throw new TypeError('Output renderer must be a function.');
    }

    this.console = cli;
    this.renderer = renderer;
  }

  /**
   * Enqueue a render. Optionally mark the update as the final render.
   */
  enqueue(final: boolean = false): this {
    if (this.isComplete()) {
      return this;
    }

    if (final) {
      this.final = true;
    }

    this.console.render(this);

    return this;
  }

  /**
   * Erase the previous content if it exists.
   */
  erasePrevious(): this {
    if (this.previousHeight > 0) {
      this.console.out(ansiEscapes.eraseLines(this.previousHeight));
    }

    return this;
  }

  /**
   * Return true if the output is complete and fully rendered.
   */
  isComplete(): boolean {
    return this.completed;
  }

  /**
   * Return true if the next render is the final render.
   */
  isFinal(): boolean {
    return this.final;
  }

  /**
   * Render the content to the console and calculate a new height.
   * Since an output represents an exact line, or a collection of lines,
   * we must always end with a new line to calculate height correctly.
   */
  render(): this {
    if (this.isComplete()) {
      return this;
    }

    let content = this.renderer();

    // Always end with a new line
    if (!content.endsWith('\n')) {
      content += '\n';
    }

    // Content cannot be higher than the terminal
    const lines = content.split('\n');
    const maxHeight = cliSize().rows - 1; // Buffer for input line

    if (lines.length >= maxHeight) {
      content = lines.slice(-maxHeight).join('\n');
    }

    // Write output
    this.console.out(content);

    // Mark output as complete if the final render
    if (this.isFinal()) {
      this.completed = true;
      // Otherwise calculate the height of the output
    } else {
      this.previousHeight = lines.length;
    }

    return this;
  }
}
