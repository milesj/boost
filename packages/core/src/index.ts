/**
 * @copyright   2017-2019, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import CLI from './CLI';
import ConfigLoader from './ConfigLoader';
import Console from './Console';
import Context from './Context';
import Emitter, { EventArguments, EventListener } from './Emitter';
import ModuleLoader from './ModuleLoader';
import Pipeline from './Pipeline';
import Plugin from './Plugin';
import Reporter from './Reporter';
import Routine from './Routine';
import Task from './Task';
import Tool, { ToolOptions, ToolConfig, ToolPluginRegistry } from './Tool';

export {
  CLI,
  ConfigLoader,
  Console,
  Context,
  Emitter,
  EventArguments,
  EventListener,
  ModuleLoader,
  Pipeline,
  Plugin,
  Reporter,
  Routine,
  Task,
  Tool,
  ToolConfig,
  ToolOptions,
  ToolPluginRegistry,
};

export * from './types';
