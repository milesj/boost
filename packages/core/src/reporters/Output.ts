/**
 * @copyright   2019, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import ansiEscapes from 'ansi-escapes';
import Console from '../Console';

export default class Output {
  console: Console;

  content: string = '';

  lastHeight: number = 0;

  constructor(console: Console) {
    this.console = console;
  }

  /**
   * Update the output block with the defined content and render the console.
   * Since an output represents an exact line, or a collection of lines,
   * we must always end with a new line to calculate height correctly.
   */
  update(content: string) {
    this.content = content;

    // Always end with a new line
    if (!this.content.endsWith('\n')) {
      this.content += '\n';
    }

    this.console.render();
  }

  /**
   * Render the content to the console. If this output has previously
   * been rendered, re-render it by erasing previous output.
   */
  render(): string {
    let output = ansiEscapes.eraseLines(this.lastHeight);

    // Count the height of the content without the ANSI escape codes above
    this.lastHeight = this.content.split('\n').length;

    output += this.content;

    return output;
  }
}
