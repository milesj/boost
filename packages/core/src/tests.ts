import path from 'path';
import { Arguments } from 'yargs-parser';
import Console from './Console';
import Context from './Context';
import Plugin from './Plugin';
import Routine from './Routine';
import Task, { TaskAction } from './Task';
import Tool, { ToolConfig, ToolOptions, ToolPluginRegistry } from './Tool';
import { Debugger, PackageConfig, PluginSetting } from './types';

export interface TestToolPlugins extends ToolPluginRegistry {
  plugin: Plugin;
}

export interface TestToolConfig extends ToolConfig {
  plugins: PluginSetting<Plugin>;
}

export type TestTool = Tool<TestToolPlugins, TestToolConfig>;

export function stubArgs(fields: object = {}): Arguments {
  return {
    $0: '',
    _: [],
    ...fields,
  };
}

export function stubPackageJson(fields: Partial<PackageConfig> = {}): PackageConfig {
  return {
    name: 'test-boost',
    version: '0.0.0',
    ...fields,
  };
}

export function stubToolConfig(config: Partial<TestToolConfig> = {}): TestToolConfig {
  return {
    debug: false,
    extends: [],
    locale: '',
    output: 2,
    plugins: [],
    reporters: [],
    settings: {},
    silent: false,
    theme: 'default',
    ...config,
  };
}

export function mockDebugger(): Debugger {
  const debug = jest.fn();

  // @ts-ignore Allow property access
  debug.invariant = jest.fn();

  return debug as any;
}

export function mockTool(options?: Partial<ToolOptions>): TestTool {
  const tool: TestTool = new Tool({
    appName: 'test-boost',
    // Match fixtures path
    appPath: path.join(process.cwd(), 'tests/__fixtures__/app'),
    ...options,
  });

  tool.registerPlugin('plugin', Plugin);

  // Stub out standard objects
  tool.args = stubArgs();
  tool.config = stubToolConfig();
  tool.package = stubPackageJson();

  // Stub out methods
  tool.debug = mockDebugger();
  tool.createDebugger = mockDebugger;

  // @ts-ignore Allow private access to avoid loaders
  tool.initialized = true;

  return tool;
}

export function mockConsole(tool: Tool<any, any>): Console {
  const cli = new Console(tool, {
    stderr: jest.fn(),
    stdout: jest.fn(),
  });

  cli.err = jest.fn();
  cli.out = jest.fn();

  return cli;
}

export class MockRoutine<Ctx extends Context> extends Routine<Ctx, TestTool> {
  execute(context: Ctx, value: unknown) {
    return Promise.resolve(value);
  }
}

export function mockRoutine<Ctx extends Context>(
  tool: Tool<any, any>,
  key: string = 'key',
  title: string = 'Title',
): Routine<Ctx, TestTool> {
  const routine = new MockRoutine<Ctx>(key, title);

  routine.tool = tool;
  routine.debug = mockDebugger();

  return routine;
}

export function mockTask<Ctx extends Context>(
  action: TaskAction<Ctx> | null = null,
  title: string = 'Description',
): Task<Ctx> {
  return new Task(title, action || jest.fn());
}
