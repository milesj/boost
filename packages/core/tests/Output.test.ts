import ansiEscapes from 'ansi-escapes';
import { mockConsole, mockTool } from '../src/tests';
import Console from '../src/Console';
import Output from '../src/Output';

jest.mock('term-size', () => () => ({ columns: 0, rows: 5 }));

describe('Output', () => {
  let output: Output;
  let cli: Console;

  beforeEach(() => {
    cli = mockConsole(mockTool());
    cli.render = jest.fn();

    output = new Output(cli, () => 'foo\nbar\nbaz');
  });

  describe('constructor()', () => {
    it('errors if renderer is not a function', () => {
      // @ts-ignore Allow invalid type
      expect(() => new Output(cli, 123)).toThrowErrorMatchingSnapshot();
    });
  });

  describe('concurrent()', () => {
    it('marks the output as concurrent', () => {
      expect(output.isConcurrent()).toBe(false);

      output.concurrent();

      expect(output.isConcurrent()).toBe(true);
    });
  });

  describe('enqueue()', () => {
    it('renders the output into the console', () => {
      output.enqueue();

      expect(cli.render).toHaveBeenCalledWith(output);
    });

    it('doesnt render the output into the console if already completed', () => {
      // @ts-ignore Allow private access
      output.state.completed = true;
      output.enqueue();

      expect(cli.render).not.toHaveBeenCalled();
    });

    it('marks the next render as the final one', () => {
      expect(output.isFinal()).toBe(false);

      output.enqueue(true);

      expect(output.isFinal()).toBe(true);
    });

    it('triggers start lifecycle', () => {
      const spy = jest.spyOn(output, 'onStart' as any);

      output.enqueue();

      expect(spy).toHaveBeenCalled();
    });

    it('doesnt trigger start lifecycle when final render', () => {
      const spy = jest.spyOn(output, 'onStart' as any);

      output.enqueue(true);

      expect(spy).not.toHaveBeenCalled();
    });

    it('triggers last lifecycle when marked as final', () => {
      const spy = jest.spyOn(output, 'onLast' as any);

      output.enqueue(true);

      expect(spy).toHaveBeenCalled();
    });

    it('doesnt trigger last lifecycle on normal enqueue', () => {
      const spy = jest.spyOn(output, 'onLast' as any);

      output.enqueue();

      expect(spy).not.toHaveBeenCalled();
    });
  });

  describe('erasePrevious()', () => {
    it('does nothing if no previous content', () => {
      output.erasePrevious();

      expect(cli.out).not.toHaveBeenCalled();
    });

    it('erases previous content based on height', () => {
      output.render();
      output.erasePrevious();

      expect(cli.out).toHaveBeenCalledWith('foo\nbar\nbaz\n');
      expect(cli.out).toHaveBeenCalledWith(ansiEscapes.eraseLines(4));
    });
  });

  describe('render()', () => {
    it('appends a trailing newline if one does not exist', () => {
      output.render();

      expect(cli.out).toHaveBeenCalledWith('foo\nbar\nbaz\n');
    });

    it('doesnt append a trailing newline if one does exist', () => {
      output = new Output(cli, () => 'foo\n');
      output.render();

      expect(cli.out).toHaveBeenCalledWith('foo\n');
    });

    it('sets previous line height if not final', () => {
      output.render();

      // @ts-ignore Allow protected access
      expect(output.previousHeight).toBe(4);
    });

    it('doesnt set previous line height if final', () => {
      output.enqueue(true);
      output.render();

      // @ts-ignore Allow protected access
      expect(output.previousHeight).toBe(0);
    });

    it('sets completed if final', () => {
      output.enqueue(true);
      output.render();

      expect(output.isComplete()).toBe(true);
    });

    it('doesnt set completed if not final', () => {
      output.render();

      expect(output.isComplete()).toBe(false);
    });

    it('truncates output if the height exceeds the terminal size', () => {
      output = new Output(cli, () => '1\n2\n3\n4\n5\n6\n7\n8');
      output.render();

      expect(cli.out).toHaveBeenCalledWith('6\n7\n8\n');
    });

    it('triggers first lifecycle on first render', () => {
      const spy = jest.spyOn(output, 'onFirst' as any);

      output.render();

      expect(spy).toHaveBeenCalled();
    });

    it('doesnt trigger first lifecycle on subsequent renders', () => {
      const spy = jest.spyOn(output, 'onFirst' as any);

      output.render();
      output.render();
      output.render();

      expect(spy).toHaveBeenCalledTimes(1);
    });

    it('triggers complete lifecycle on last render when marked as final', () => {
      const spy = jest.spyOn(output, 'onComplete' as any);

      output.enqueue(true);
      output.render();

      expect(spy).toHaveBeenCalled();
    });

    it('doesnt trigger complete lifecycle when not final', () => {
      const spy = jest.spyOn(output, 'onComplete' as any);

      output.render();

      expect(spy).not.toHaveBeenCalled();
    });
  });
});
