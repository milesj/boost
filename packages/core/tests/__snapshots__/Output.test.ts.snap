// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Output constructor() errors if renderer is not a function 1`] = `"Output renderer must be a function."`;
